File: doc/src_tutorial/mxm.f90
program main
	parameter(msize=1000)
	real(kind=8), allocatable :: a(:,:),b(:,:),c(:,:)
	
	allocate (a(msize,msize), b(msize,msize), c(msize,msize), stat=istat)
	if (istat.ne.0) then
	    stop "*** Allocate() failed."
	endif
	n=msize
	
    !!	ninit=10
    !!	call f_pm_initialize (ninit)
	
    !!	call f_pm_start ("A:subinit2d")
	call subinit2d (msize,n,a,b,c)
    !!	call f_pm_stop ("A:subinit2d")
	
    !!	call f_pm_start ("B:submxm2d")
	call submxm2d (msize,n,dflop,a,b,c)
    !!	call f_pm_stop ("B:submxm2d")
	
    !!	call f_pm_report ("")
	write(*,*)  'something was computed', c(msize,msize)
	
	stop
end program
	
subroutine subinit2d (msize,n,a,b,c)
	real(kind=8) :: a(msize,msize), b(msize,msize), c(msize,msize)
	!!  !$omp parallel do 
	do j=1, n
	do i=1, n
	a(i,j)=sin(real(i)/real(n))
	b(i,j)=cos(real(i)/real(n))
	end do
	end do
	return
end subroutine
	
subroutine submxm2d (msize,n,dflop,a,b,c)
	real(kind=8) :: a(msize,msize), b(msize,msize), c(msize,msize)
	real(kind=8) :: x
	
	dflop=2.0*dble(n)**3
	!!  !$omp parallel do private(i,j,k,x)
	do j=1, n
	do i=1, n
	x=0
	do k=1, n
	x=x+a(i,k)*b(k,j)
	end do
	c(i,j) = x
	end do
	end do
	return
end subroutine

File: doc/src_tutorial/mxm.cpp
#include <string>
#include <math.h>
#include <stdio.h>
void init2d();
void mxm2d();
	//	#include <PerfMonitor.h>
	//	using namespace pm_lib;
	//	PerfMonitor PM;

#define MATSIZE 1000
int nsize;
struct matrix {
	int nsize;
	float a2[MATSIZE][MATSIZE];
	float b2[MATSIZE][MATSIZE];
	float c2[MATSIZE][MATSIZE];
} matrix;

int main (int argc, char *argv[])
{
	//	PM.initialize();
	//	PM.start("A:init2d");
	init2d();
	//	PM.stop ("A:init2d");
	//	PM.start("B:mxm2d");
	mxm2d();
	//	PM.stop ("B:mxm2d");
	//	PM.print(stdout, "", "", 0);
	//	PM.printDetail(stdout);
	//	PM.printThreads(stdout, 0, 0);
	printf("something was computed... %f\n",c2[0][0]);
	return 0;
}


void init2d()
{
	int i, j, nsize;
	matrix.nsize = MATSIZE;
	nsize = matrix.nsize;
	//	//	#pragma omp parallel for private(i,j)
	for (i=0; i<nsize; i++){
	for (j=0; j<nsize; j++){
	matrix.a2[i][j] = sin((float)j/(float)nsize);
	matrix.b2[i][j] = cos((float)j/(float)nsize);
	matrix.c2[i][j] = 0.0;
	}
	}
}

void mxm2d()
{
	int i, j, k, nsize;
	float c1,c2,c3;
	nsize = matrix.nsize;
	//	//	#pragma omp parallel for private(i,j,k,c1)
	for (i=0; i<nsize; i++){
	for (j=0; j<nsize; j++){
		c1=0.0;
		for (k=0; k<nsize; k++){
		c1=c1 + matrix.a2[k][i] * matrix.b2[j][k];
		}
		matrix.c2[j][i] = c1;
	}
	}
}


